benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,228 @@
+"""Classes for persisting benchmark results."""
+
+import json
+import os
+from datetime import datetime
+from enum import Enum
+from pathlib import Path
+from typing import Protocol
+
+import boto3
+import boto3.session
+from attr import define, field
+from attrs.validators import instance_of
+from boto3.session import Session
+from git import Repo
+
+from benchmarks import Benchmark, Result
+
+
+class PathStrategy(Enum):
+    """The way a path extension is constructed."""
+
+    HIERARCHICAL = 1
+    """Refers to the hierarchical path construction of folders."""
+
+    FLAT = 2
+    """Refers to the flat path construction of a single file name."""
+
+
+@define
+class PathConstructor:
+    """A class to construct the hierarchical path of a result object.
+
+    This class is used to encapsulate the construction of a respective path depending
+    on where the object should be stored. Since different storage backends have
+    different requirements for the path, and some like the used S3 Bucket
+    not even support folders, the path uses its variables to construct the path
+    in the order of the variables with a separator depending on the strategy.
+    For compatibility reasons, the path is sanitized to only contain the following
+    allowed characters:
+    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-."
+    """
+
+    benchmark_name: str = field(validator=instance_of(str))
+    """The name of the benchmark for which the path should be constructed."""
+
+    branch: str = field(validator=instance_of(str), init=False)
+    """The branch currently checked out."""
+
+    latest_baybe_tag: str = field(validator=instance_of(str))
+    """The latest BayBE version tag."""
+
+    execution_date_time: datetime = field(validator=instance_of(datetime))
+    """The date and time when the benchmark was executed."""
+
+    commit_hash: str = field(validator=instance_of(str))
+    """The latest commit hash of the current checkout branch."""
```

Comment:
I guess you mean `The hash of the commit currently checked out`? Because the your description could mean something else in case there are later commits in the same branch